File: cpp/sparsearray.cpp
#include "sparsearray.h"
#include <stddef.h>
#include <node.h>

/**
 * Constructor for a new sparse array
 */
Sparsearray::Sparsearray(int w, int h, int dl, unsigned int *dta) {
  width = w;
  height = h;
  datalen = dl;
  data = dta;
}

/**
 * Constructor for a new sparse array
 */
Sparsearray::Sparsearray() {
  width = 0;
  height = 0;
  datalen = 0;
  data = NULL;
}

/**
 * Destructor for a new sparse array
 */
Sparsearray::~Sparsearray() {
  delete [] data;
  data = NULL;
}

/**
 * Get the area of the array
 */
unsigned int Sparsearray::area() {
  return width * height;
}

File: lib/sparsearray.js
/**
 * Created by whitehawk on 2015-11-26.
 */

var SparseArray = function (width, height, dataarray) {
    var instructions = "Constructor for SparseArray. Eg: var a = new SparseArray(width, height, data);\n\n";
    this.width = width;
    this.height = height;
    this.data = dataarray;
    if (!width || !height || !(dataarray.length)) {
        throw new Error(instructions + "Required arguments: width (Number), height (Number), and dataarray (Array).");
    }
    if (width < 1 || height < 1) {
        throw new Error(instructions + "Width and height must be non-zero.");
    }
    if (dataarray.length % 3 !== 0) {
        throw new Error(instructions + "The data array must be an ordered list of UInt32's in the format of: n, x, y.");
    }
    if (dataarray.length > 0 && typeof(dataarray[0]) != 'number') {
        throw new Error("The data array must contain Numbers. Seems to contain '" + typeof(dataarray[0]) + "'.");
    }
};

// Export it
module.exports = SparseArray;